
<!-- dynadb_Dynamics.html -->
<head><style>
.long input {
 width:100%;
}

.deco a:link, .deco a:visited {
    background-color: #F0F0F0;
    color: #337ac1;
    text-align: center;
    text-decoration: none;
    border: 1px solid #337ac1;
    border-radius: 12px;
}
.deco a:hover, .deco a:active {
    background-color: #337ac1;
    color:white;
}
</style></head>

{% extends "home/base.html" %} 
{% load static %} 


{% block content %}

<form method="post" id="myform" action="{{ action }}" enctype="multipart/form-data"> 


 <!-- description column -->
 <div class="col-md-4">
     
     <div class="col-md-12 panel panel-primary" style="background-color:#3278B4;text-align:center;color:white;">   
         <div class="panel-body">
        	<h3 > Simulation Submission<br>Step 3:<br>Complex Information</h3>
         </div> 
     </div>

     <!-- main description -->
     <div class="col-md-12 panel panel-primary">
         <div class="panel-body">
            <div style="text-align:center"> 
             <h4>FILL IN THE COMPLEX FORM</h4>
            </div>
             <p style="text-align: justify;" >Fill in the form with information on the curated pdb protein (loops, mutations, etc)</p>
             <p style="text-align: justify;" >Contains structural information about proteins or “molecular complexes” in PDB format similar to the PDB database: structure and coordinates of protein, ligand and/or  molecules crystallized with proteins, crystallized or user-predicted stable waters and ions.</p>

             <p style="text-align: justify;">Note, do not include coordinates of solvent, ions, membrane lipids or other molecules which are part of the general solvation environment specially if their coordinates are randomly generated. </p>
         </div>
     </div>

     <!-- link to docs -->
     <div class="col-md-12 panel panel-primary">
         <div class="panel-body">
             <p>For more information on this form, see the <a alt="Work in progress.">docs</a>.
             </p>
         </div>
     </div>
  </div>
 <!--   _____________________________________________________________________________________________-->

<!-- middle column -->
<div class="col-md-7 "> 
    <!-- filters -->
  <div class="col-lg-12 panel panel-primary"> 

    <div class="panel-body">
 <!------------------------------------------------------------------------------ middle column -->
<!--        <h4> <a title="Table containing molecular dynamic metadata"> dyndb_Dynamics Table:</a></h4> <br> -->
<!--	<div style=display:table;>  -->

		<div class=row id='PRUEBA'>
<!-- 			<div style="text-align:center">
            			<h4><br>Select Complex</h4><br>    
             			<h4><br>Complex Information</h4><br>
            		</div>   -->

 			<div style="text-align:center">
<!--       			<h4><br><a title="El antiguo html esta comentado en las siguientes lineas del fichero MODEL.html" >Reuse Complex Temporarily unavailable</a></h4><br> -->
             			<h4><br>Complex Information</h4><br>
<!--             			<h4><br><a title="El antiguo html esta comentado en las siguientes lineas del fichero MODEL.html" >Complex Information</a></h4><br> -->
            		</div>
                <div class="row">
		  	<div class="col-md-1"></div><div class="col-md-3" style="text-align:left"><label for="Model">Complex:</label></div>
		  	<div class="col-md-4" style="text-align:left"> Reuse Complex <input  type="radio" name="Model" checked value="Reuse Model" disabled></div>
	 		<div class="col-md-4" style="color:gainsboro;text-align:center"><input  type="radio" name="Model" value="New Model" disabled > New Complex</div></div><br>
	        <div class="row">
			<div class="col-md-8 col-md-offset-4" style="text-align:left">
			    <div class="row">
				<div class="col-md-4" style="text-align:left;height:30px;margin: 0px 0px 5px 0px;"><label for="Choose_reused_model"><a title="Choose a previous complex from the GPCRmd database">Complex ID:</a> </label></div>
				<div class="col-md-8"><input style="width:130px" id="Choose_reused_model" name="Choose_reused_model" rows="1" placeholder="Previous Complex" ></input></div>
                            </div>
			    <div class="row">
				<div class="col-md-4" style="text-align:left;height:30px;margin: 5px 0px px 0px;"><label for="Choose_submission_id"><a title="Choose a previous submission from the GPCRmd database">Submission ID:</a> </label></div>

			        <div class="col-md-8"><input style="width:130px" id="Choose_submission_id" name="Choose_submission_id" rows="1" placeholder="Previous Complex" ></input></div>
		            </div><br>
                            </div>
				
		<br>
		</div></div> 
		</div></div> 
    <!-- filters -->
  <div class="col-lg-12 panel panel-primary"> 
  <div class="row"> 
     <br><br>
	        <div class="row">
			<div class="col-md-1"style="height:30px"></div><div class="col-md-3"><label for="id_name"><a title="Choose a name for your complex" > Complex Name:</a></label></div>
			<div class="col-md-6"><input cols="40" id="id_name" type="text" name="name" rows="2" placeholder="5-HT1B receptor in complex with Clozapine"  style="width:300px"></input></div>
		<br></div>

	        <div class="row">
			<div class="col-md-1"style="height:30px"></div><div class="col-md-3"><label for="id_type"><a title="Complex Type: Type of structure that is modeled." > Complex Type:</a></label></div>
				<div class="col-md-6" style="height:30px"><select id="id_type" name="type" placeholder="   "  style="width:300px" >
					<option value="0">Protein Apoform </option>
					<option value="1"> Protein-Protein or Protein-Ligand Complex</option>
				</select></div>
		<br></div>

	        <div class="row">
			<div class="col-md-1"style="height:30px"></div><div class="col-md-3"><label for="id_pdbid"><a title="PDB ID: PDB ID from Protein Data Bank database. It should contain chain code when necessary (e.g. 4IAR.A). " >PDB ID:</a></label></div>
			<div class="col-md-6"><input id="id_pdbid" maxlength="6" name="pdbid" type="text" placeholder=" 4IAR.A"  style="width:300px"/></div>
		<br></div>

	        <div class="row">
		<div class="col-md-1"style="height:30px"></div><div class="col-md-3"><label for="id_description"><a title="Description: Description of the complex containing information that is not provided in other field such as procedure followed for the construction of the model. " >Description:</a></label></div>
			<div class="col-md-6"><textarea cols="40" id="id_description" name="description" rows="2" placeholder=" Agonist conformation (ergotamine). Clozapine binding by docking. Loop repair in Modeller 9v10."  style="width:300px"></textarea></div>
		<br></div>

	        <div class="row">
			<div class="col-md-1"style="height:30px"></div><div class="col-md-3"><label for="id_source_type"><a title="Source Type: Source of the coordinates of the “complex”. Loop repairs and other replaced missing protein residues on X-ray or NMR structures are not taken into account for setting the value of this field. In order to set the value to X-ray or NMR all the coordinates (including relative positions between different components) of protein complexes and their components should come from an X-ray or an NMR experiment." >Source Type:</a></label></div>
				<div class="col-md-6" style="height:30px"><select id="id_source_type" name="source_type" placeholder="   "  style="width:300px" >
					<option value="0">X-ray</option>
					<option value="1">NMR</option>
					<option value="2">Docking</option>
					<option value="3">MD</option>
					<option value="4">Other</option>
				</select></div>
		<br></div>


	        <div class="row">
                	<div class="col-md-1"style="height:30px"></div><div class="col-md-3"><label for="upload_pdb"><a title="Upload PDB file: Provide the PDB file which contains the 3D structure of the model (Max. size of the file is 50 MB)" >Upload PDB file:</a></label></div>
			<div class="col-md-3"><input type="file"  name="file_source"  class="validate[required,custom[onlyLetter],length[0,100]] feedback-input" placeholder=" Coordinates (PDB)" id="PDBcoor" /></div>
		</div>

			<div style="text-align:right;margin-right:5%"><input type="submit" value="Upload" name="upload`_pdb"  class="validate[required,custom[onlyLetter],length[0,100]] feedback-input" id="PDBcoor" /></div>
			
  <br></div>
		<div class=row id='PRUEBA1'>
 			<div style="text-align:center">

             			<h4>Residues assignment and protein coordinate sources</h4><br>
                        <div class="col-md-12" style="heigh:30px">
                                <p style="font-size:12px;text-align:justify;margin-left:10%;margin-right:10%">In order to assign correctly the canonical GPCR numbering to your simulation system, please match the residue ID used in your simulation system with the standard/referemce residue id of the UNIPROT sequence</p><br>
                                </div>
            		</div>


                        <table style="text-align=center;width:96%;margin-left:2%;margin-right:2%;"> 
                                <thead><tr><th style="width:40px;text-align:center"><label style="writing-mode: vertical-rl;text-align:right;height:65px"for="id_prot" size='10'><a title="Ordinal number corresponding to the records containing sequence information that was provided in the protein form ">Prot</a></label></th>
					<!--<th style="width:70px;text-align:center"><label for="id_id_protein" size='10'><a title="Protein ID: ID of the protein that is modeled in this fragment.">Protein ID</a></label></th>-->
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_chain" size='10'><a title="Chain: Chain identifier for the model in PDB format. ">Chain</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_segid" size='10'><a title="Segid: segment identifier for the model in charmm PDB format. ">Segid</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right" for="id_resid_from" size='10'><a title="Res from: Starting PDB file resid of the protein fragment. ">Res from</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_resid_to"><a title="Resid to: ending PDB file resid of the protein fragment.">Res to</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_seq_resid_from"><a title="Seq Resid to: Starting resid of the protein fragment in the protein sequence numbering."> Seq Res from</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_seq_resid_to"><a title=" Seq Resid to: ending resid of the protein fragment in the protein sequence numbering."> Seq Res to</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_id_bonded_to_previous"><a title="Bond?: check this box if the fragment is covalently bonded to the C-terminal of the prevous one."> Bond?</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_pdbidps"><a title="PDB ID: PDB ID from Protein Data Bank database. It should contain chain code when necessary (e.g. 4IAR.A). Must be empty when the source of the modeled residues is not X-ray or NMR. ">PDB ID</a></label></th>
					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_source_typeps"><a title="Source type: Source of the coordinates of the protein fragment.">Source type</a></label></th>
<!--					<th style="width:70px;text-align:center"><label style="writing-mode: vertical-rl;height:65px;text-align:right"for="id_template_id_model"><a title="Template id model: ID of the model used as template if “Source type” is set to 'Homology'. ">Template model ID</a></label></th></tr> -->
                                </thead>
                                <tbody id="pElement1">
{% for o in rows %}           <table> </tbody> 
				<tr id="Element1"><td style="padding:5px 5px 5px 5px"><input id="id_prot" value="1"  name="prot" type="number" style="width:45px" placeholder="  1" /></td>
				<td style="padding:5px 5px 5px 5px"><input id="id_id_protein" value=  name="id_protein" type="number" style="width:60px" placeholder=" 1"/></td>
				<td style="padding:5px 5px 5px 5px"><input id="id_chain" maxlength="1" value={{o.chain}}  name="chain" type="text" style="width:45px" placeholder=" E"/></td>
				<td style="padding:5px 5px 5px 5px"><input id="id_segid" maxlength="4" value={{o.segid}}  name="segid" type="text" style="width:45px" placeholder="PROA"/></td>
				<td style="padding:5px 5px 5px 5px"><input id="id_resid_from" value={{o.resid_from}}  name="resid_from" type="text"  style="width:45px" placeholder=" "/></td>
				<td style="padding:5px 5px 5px 5px"><input id="id_resid_to" value={{o.resid_to}}  name="resid_to" type="text" style="width:50px" placeholder=" ?? "/></td>
                              <td style="padding:5px 5px 5px 5px"><input id="id_seq_resid_from" value={{o.seq_resid_from}}  name="seq_resid_from" type="text"  style="width:50px" placeholder=" "/></td>
                              <td style="padding:5px 5px 5px 5px"><input id="id_seq_resid_to" value={{o.seq_resid_to}}  name="seq_resid_to" type="text"  style="width:50px" placeholder=" "/></td>
                              <td style="padding:5px 5px 5px 10px"><input id="id_bonded_to_id_modeled_residues" value={{o.bonded_to_id_modeled_residues_id}}  name="bonded_to_id_modeled_residues" style="aligned:center" type="checkbox" checked placeholder=" "/></td>
				<td style="padding:5px 5px 5px 5px"><input id="id_pdbidps" maxlength="6" value={{o.pdbid}}  name="pdbidps" type="text"  style="width:50px" placeholder=" 4IAR.A  "/></td>
				<td style="padding:5px 5px 5px 5px"><select id="id_source_typeps"  value={{o.source_type}}  name="source_typeps" style="width:50px" >	
					<option value="0">X-ray</option>
					<option value="1">NMR</option>
					<option value="2">Ab-initio</option>
					<option value="3">Threading</option>
					<option value="4">Homology</option>
					<option value="5">Threading</option>
					<option value="6">Other Computational Methods</option>
				</select></td>
<!--				<td style="padding:5px 5px 5px 5px"><input id="id_template_id_model" style="width:60px"  value= name="template_id_model" type=text></td>   -->
				</tr></tbody></table> 



{% endfor %}

{% endblock content %}

